### PR TITLE
Relicensing IE!

### PR DESCRIPTION
**Pull request by BluSunrize**
Righto!

Here goes. I want to (finally) give IE a proper license, rather than "All Rights Reserved with me making weird exceptions".
I wrote this custom license and tested it on Trauma, I ran it past multiple programmer friends and arrive at what it is now, I do hope it covers everything. (Should I have missed something, please let me know so I can fix it before merging this)

I wish to relicense all of IE's code and assets to this license. However, I require the approval of all contributors.
IE's current codebase consists mainly of my own work, the amazing contributions of @malte0811 and the 1.11 port work of @AtomicBlom.
Of you two, I would like an explicit statement if you approve of this relicensing or if you wish to have anything changed / withdraw any of your code beforehand.

Anyone else who ever contributed to IE:
I will leave this open for a whole week, until October 15th. On that day, I will consider anyone who didn't oppose this to be in favour, and any claims made after that are considered null and void.

### CODE REVIEW

**Comment by malte0811** [comment]
I approve of the new license, esp. the last bit of line 18 since I would have to rewrite a lot of code for IW otherwise.
>I will leave this open for a whole week, until October 15th. On that day, I will consider anyone who didn't oppose this to be in favour, and any claims made after that are considered null and void.

I'm not sure that's legally possible. When Forge got relicensed they asked for permission from everyone who ever contributed. (But then, I doubt that this license would actually be valid/enforceable legally)

@FerroO2000 [this](https://github.com/BluSunrize/ImmersiveEngineering/blob/d9f2a6034e957a4ab070095cb73c01f21751f873/LICENSE#L18) may be relevant to you, last time I checked your generic block classes were direct copies of IE's without anything noting that they were.

**Comment by BluSunrize** [comment]
@3TUSK @xaxa123 @xbony2 @alexbegt @Landmaster @theunkn0wn1 @primetoxinz @TheGuyRunningSouth @mconwa01 @MatrixN1 @SkySom @MalkContent @mezz @Drullkus @GWYOG @Faxn @berinhardt @InsomniaKitten @warlordjones @PaleoCrafter @dogboy21

Ping for literally everyone who ever contributed

**Comment by FerroO2000** [comment]
Sorry, I will write the provenance. Let me only a couple of day

**Comment by BluSunrize** [comment]
All any of you need to do really is to say "This is fine, do whatever".
This relicensing is nothing harmful, it just makes it easier for addon devs in the future, and forks and pullrequests. Basically saves me all the people asking "Can I do this?"

**Comment by primetoxinz** [comment]
Fine by me

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+#### What you *can not* do:
+- Claim any of this code to be your property
+- Redistribute this project (source or compiled) without explicit permission of the license holder, BluSunrize
+- Copy large sections of this code (classes, complex methods) without giving credit
+- Monetize any of this project. This means selling the code, compiled projects and also forbids the use of paywalls or "donations" to restrict access to any part of the project's content. I specifically exempt CurseForge's "Points"-system from this statement.
+
+#### What you *have to* abide by:
+- If you have used any code from this project (with the exception of API classes), that project must also be visible source. This allows others to learn from your work.
+- Give proper credit where it's due when copying copy (see above)
```
Comment (PaleoCrafter) [inline LICENSE]: `when copying copy` - This part reads a little weirdly, maybe you meant `when copying code`?

**Comment by theunkn0wn1** [comment]
I have reviewed the license as described in commit d9f2a6034e957a4ab070095cb73c01f21751f873.
I have no objections, glad to see you have a proper license file now.

File: LICENSE [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+#### On the topic of *resources*:
+- Always give credit to original artists
+- Always ask for permission before copying assets
+- Addons mods are allowed to use and remix art assets for the sole purpose of being addons. When in doubt, better ask.
```
Comment (PaleoCrafter) [inline LICENSE]: Minor typo: `Addons mods`, should probably read `Addon mods`

**Comment by mezz** [comment]
If you're serious you should probably use a known license like LGPLv3 (Forestry uses that). A custom license like this is not going to hold up if you actually need to use it.
Personally I think licensing in modded minecraft is already baseless so I wont object to you re-licensing, do what you want :)

File: LICENSE (hunk 2 above)
Comment (InsomniaKitten) [reply LICENSE]: I personally think it would be clearer to say `When in doubt, just ask.`

**Comment by InsomniaKitten** [comment]
I'm fine with this ^^

File: LICENSE (hunk 1 above)
Comment (theunkn0wn1) [reply LICENSE]: I agree, `when copying copy` makes no sense. One way to resolve this is to `...when making copies` or simply `when copying portions of the code`

**Comment by dogboy21** [comment]
I'm also fine with the new licence

**Comment by FerroO2000** [comment]
@BluSunrize ok, thanks

File: LICENSE [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+#### Contributing to this repository:
+- The original author keeps copyright of their contibuted code and resources
+- The owner of this repository is given rights to usage and modification of the contributed code or resource. This may not be withdrawn
+- Any code or resource submitted must be licensed to this project with this license
```
Comment (3TUSK) [inline LICENSE]: I suggest use CLA for this.

File: LICENSE [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+- Claim any of this code to be your property
+- Redistribute this project (source or compiled) without explicit permission of the license holder, BluSunrize
+- Copy large sections of this code (classes, complex methods) without giving credit
+- Monetize any of this project. This means selling the code, compiled projects and also forbids the use of paywalls or "donations" to restrict access to any part of the project's content. I specifically exempt CurseForge's "Points"-system from this statement.
+
+#### What you *have to* abide by:
+- If you have used any code from this project (with the exception of API classes), that project must also be visible source. This allows others to learn from your work.
+- Give proper credit where it's due when copying copy (see above)
+
+#### On the topic of *resources*:
```
Comment (3TUSK) [inline LICENSE]: This sections only covers assets of art work. Is that intended?

File: LICENSE (hunk 3 above)
Comment (BluSunrize) [reply LICENSE]: I dunno how that works :V

**Comment by 3TUSK** [comment]
From my interpretation of both "What you *can* do" and "What you *can not* do" sections, I agree with @mezz on using [a license that approved by OSI](https://opensource.org/licenses), to be specific, I believe LGPLv2.1 or later version can cover your use case. 

You can also use separate license for your API, your main code, your assets and even the text (see how [Applied Energistics 2](https://github.com/AppliedEnergistics/Applied-Energistics-2) and [Charset](https://github.com/CharsetMC/Charset/tree/1.12/docs/licenses) handles license).

Regarding re-license: consider my contribution as in public domain.

File: LICENSE (hunk 4 above)
Comment (BluSunrize) [reply LICENSE]: tbh, I didn't really *think* about language files at the time, and art+models are really the only resources that mattered in that regard .-.

File: LICENSE (hunk 3 above)
Comment (3TUSK) [reply LICENSE]: https://cla-assistant.io/ <- I think you can find some "getting started" guide here?

File: LICENSE (hunk 4 above)
Comment (3TUSK) [reply LICENSE]: I was also thinking model json files and recipe json files as well -_-

**Comment by MalkContent** [comment]
Do whatever. I'm refuse to be part of licensing until I did more than fix a typo ^^

File: LICENSE (hunk 1 above)
Comment (BluSunrize) [reply LICENSE]: should have been "when copying code" will fix =P

**Comment by xbony2** [comment]
You might be missing contributions since you broke the git repository long ago. Most of contributions were from the earlier repo.

Consider my contributions to be in the public domain.

**Comment by Drullkus** [comment]
I made assets specifically for you and whatever domain you have chosen for it.

In short words, I'm totally chill about it, bro.

**Comment by GWYOG** [comment]
The new license is fine with me, do whatever you want :)

**Comment by malte0811** [comment]
Pinging everyone who contributed in 1.7.10 (unless I know they were already pinged) since they aren't in the git history for `master` any more (as @xbony2 noted).
@cobra
@Jaccob
@karakufire
@bakaxyf
@UnwrittenFun
@LYDFalse
@marcin212
@3TUSK
@Yulife
@Adaptivity
@Glassmaker
@Remeh84
@mdfntr
@Haggle1996
@artdude543
Mary (Doesn't seem to have a github account (any more?))
@Mazdallier
@berinhardt
@svenstaro
@TheJulianJES
@XFactHD
@wafflemans
@webmilio

**Comment by TheJulianJES** [comment]
I'm fine with it.

**Comment by Faxn** [comment]
I approve of, and will allow this. My contributions are minor anyway.

**Comment by svenstaro** [comment]
Fine.

**Comment by mconwa01** [comment]
Fine by me.

**Comment by Adaptivity** [comment]
Whatever changes I made belong to Immersive Engineering devs, you can relicense it without asking.

**Comment by LYDfalse** [comment]
Fine by me.

**Comment by warlordjones** [comment]
I don't even remember what I contributed. Go right ahead.

**Comment by Haggle1996** [comment]
Yes, I consider my pull request to be in the public domain. Use it as you please.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by berinhardt** [comment]
Seems reasonable to me

**Comment by SkySom** [comment]
Do with it as you please.

**Comment by cobra** [comment]
No objections from me.

**Comment by webmilio** [comment]
Its cool.

**Comment by AtomicBlom** [comment]
You have my permission.
And my axe.
And my bow.

**Comment by Landmaster** [comment]
I agree with comments by @mezz and @3TUSK that a well-known license (e.g. LGPL v3) may be preferable, but otherwise, this license is fine with me.

File: LICENSE [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+As such, this is my custom license which should hopefully cover all that:
+
+#### What you *can* do:
+- Look at the code, draw inspiration, copy bits of it to use in your own projects
+- Fork this repository and modify its code
+- Submit Pull Requests to this repository
+- Write other projects that base on this project and its API (if present) (This means Addons are totally cool)
+
+#### What you *can not* do:
```
Comment (Landmaster) [inline LICENSE]: Should be `*cannot*`

**Comment by Remeh84** [comment]
Fine by me, do whatever you want because I dont even remember what exactly I did lol

**Comment by josephcsible** [comment]
Not a contributor, so you don't have to listen to me, but I agree with @mezz and @3TUSK that you should use an OSI-approved license.

**Comment by marcin212** [comment]
You have my permission.

**Comment by artdude543** [comment]
If you continue to post the cheese @BluSunrize you can have my permission 😉 (but in serious and stuffs my permission is granted like planting a new seed)

**Comment by bakaxyf** [comment]
You have my permission.

**Comment by TheGuyRunningSouth** [comment]
Sorry I'm late, my email freaked out apparently. I'm all for visible source software, so you've got my permission.

**Comment by BluSunrize** [comment]
Well, we got most people in on this and I fixed anything marked & relevant.

I am hereby merging this and declaring IE as officially relicensed!

**Comment by 3TUSK** [comment]
> we got most people in on this

Not all of them?

**Comment by BluSunrize** [comment]
I also left it open a week longer than originally planned. There was an ultimatum and everyone was pinged. I did my dues.

**Comment by unascribed** [comment]
That's... not how licensing works. You can't relicense someone's code without their permission, period, unless you had a CLA, which, you didn't. No matter how much time you gave them to see this PR.

This is a terrible non-license. This is *worse* than just being "all-rights-reserved with weird exceptions". At least there you had legal ground, but now you're trying to define your exceptions in a legally binding way, and it does. not. work.

Please, for the love of god, switch to a known-good license. The one closest to what you're getting at with this license is the LGPLv2.1, which is also the license Forge is under. The LGPL doesn't forbid monetization, but anyone who would monetize the mod against your wishes won't be listening to this "license" either. (You may be tempted to go for the GPL, don't. It's not compatible with Minecraft.)

You can then license your assets under a separate license, the CC BY-NC-SA. It allows people to remix and redistribute the assets, as long as they credit you. There is *no* differentiation between making addons and people redistributing the mod proper, that's just not how it works. And the NC (noncommercial) clause gets you your non-monetization wish. If someone wanted to monetize a fork of your mod, they'd have to redraw every asset. And I feel like as long as I'm mentioning CC licenses, I should say **do not license the code under a CC license**. They're not designed for code.

(As long as you're mixing licenses, you probably also want to put your API under the MIT license.)

I doubt you care about anything I just said. Licensing is often misunderstood and ignored in the Minecraft community. Visible source is *broken*, because simply reading the code opens up a massive can of worms. Please, please please switch to an OSI-approved license.

**Comment by Boundarybreaker** [comment]
@unascribed makes a really good point, and the license also brings up the issue of carrying IE on when you inevitably decide to step away from modding. Both Asie and Elucent have stepped away from modding in the past week, and this license could make it so that nobody would be able to keep working on IE if/when you leave.

**Comment by 3TUSK** [comment]
What @unascribed said is exactly why I ask so. Either waiting for permission, or go ahead and do a rewrite from scratch. 

> I guess things like `System.out.println` or other sorts can't really count?

**Comment by 3TUSK** [comment]
Also, @unascribed 
> You may be tempted to go for the GPL, don't. It's not compatible with Minecraft.

It can be compatible. Have you ever noticed that FoamFix is licensed under GPL v3 (or it's v2.1? can't recall atm)? Asie added a special appendix dedicated for conveying covered work with Minecraft.

**Comment by unascribed** [comment]
Yes, I'm aware. The exception isn't foolproof, however. It's better to use the more permissive LGPL.

**Comment by BluSunrize** [comment]
@unascribed Thank you kindly for the input.
Shame that such a detailed description of how to handle the situation wasn't provided sooner. Would have probably had me reconsider the choice of the custom license.
Yeah my solution is definitely not ideal, but it's also not legally enforcable anyway, is it? So who gives a shit. I'm going to *stick* with this one for now, and for future projects I will keep the options listed above in mind.

There's other examples of mods out there that run a custom license. And they also don't have a CLA. And noone gives them shit for it, as far as I can see.

If they can fare well on custom license, why shouldn't I? In the end, this is a hobby. I don't intend to ever start some legal battle over this stuff. Should it ever get to that point, I'd probably just quit outright to save myself the hassle.

Oh, and re: someone else maintaining IE should I ever drop it: I really don't want that? It may seem egoistical to most people, but when I eventually end this project, I intend it put it at rest, and let other projects take to the stage instead.

**Comment by Boundarybreaker** [comment]
I think the difference is that most projects that have a custom license _started_ with it, instead of getting it partway through their lifespan. If you don't intend to really go ahead with any legal stuff because of the license I think it'll be ok, but definitely start out with the license at the start any future projects if you wanna use it.

**Pull request merged**